backend/assignments/2/src/main/java/com/caching/utility/ExternalAPIUtility.java

```diff
@@ -0,0 +1,118 @@
+package com.caching.utility;
+
+import com.caching.dto.GeocodeResponse;
+import com.caching.dto.ReverseGeocodeResponse;
+import com.caching.exception.CustomException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+
+/**
+ * Utility class for interacting with external geocoding APIs.
+ */
+@Component
+public class ExternalAPIUtility {
+
+    private final Logger logger = LoggerFactory.getLogger(ExternalAPIUtility.class);
+
+    private static final String ERROR = "Error";
+    @Value("${positionstack.api.key}")
+    private String positionStackApiKey;
+
+    /**
+     * Retrieves geocode information for a given address using the PositionStack API.
+     *
+     * @param address The address to geocode.
+     * @return A GeocodeResponse object containing the geocode information.
+     * @throws CustomException if there is an error during the API call.
+     */
+    public GeocodeResponse getGeocode(String address) {
+        try {
+            String encodedAddress = java.net.URLEncoder.encode(address, StandardCharsets.UTF_8.toString());
+            URL url = new URL("http://api.positionstack.com/v1/forward?access_key=" + positionStackApiKey + "&query=" + encodedAddress);
+            String response = sendRequest(url);
+
+            ObjectMapper mapper = new ObjectMapper();
+            JsonNode root = mapper.readTree(response);
+            if (root.has(ERROR)) {
+                throw new CustomException("API Error: " + root.path(ERROR).toString());
+            }
+            JsonNode firstResult = root.path("data").path(0);
+            if (firstResult.isMissingNode()) {
+                throw new CustomException("No results found");
```

Comment:
The custom exceptions could have been more specific wrt the exception they are catching..